osu.Game.Rulesets.Osu/Mods/OsuModRelax.cs

```diff
@@ -101,8 +104,14 @@ public void Update(Playfield playfield)
 
             if (requiresHold)
                 changeState(true);
-            else if (isDownState && time - lastStateChangeTime > AutoGenerator.KEY_UP_DELAY)
-                changeState(false);
+            else
+            {
+                bool hasCompleted = time > lastHitTime;
```

Comment:
Did the code run when the last frame updates?
bool hasCompleted = time > lastHitTime; fails at a 5% chance

bool hasCompleted = time >= lastHitTime; fails

bool hasCompleted = time >= lastHitTime - 1; ok, sometimes fail

I changed it from last commit because there's no reason to waste 50ms juat for a key up frame

Reply:
I'm not sure what you mean by this, nor do I understand how "fails at a 5% chance" makes sense, but theoretically, `time > lastHitTime` is supposed to be correct? if there's an issue with that, then you, as the author of this PR, should debug it and point out exactly why it's failing. "5% chance" makes no sense to me.

Reply:
It is theoretically correct, but the test sometimes fails with a extra LeftButton pressed
![image](https://user-images.githubusercontent.com/50285552/126327895-9b81fc1b-fae9-42c0-972e-3826e5ccb4ec.png)

Reply:
The reason why it sometimes fails is that relax is allowed to hit the object up to 3 miliseconds _before_ it actually starts.

https://github.com/ppy/osu/blob/738d3217882ee473fede0e47c929945fe0b08cdd/osu.Game.Rulesets.Osu/Mods/OsuModRelax.cs#L68-L70

Which is exactly what's happening on the screenshot above; `2550 - 3 < 2548.67...`. In that case the `hasCompleted` check fails, because `time` is below the last object's end time (2550), and there is no release frame emitted at the end.

A potential way to fix would be to change the check to `time > lastHitTime - relax_leniency`, but it needs to be checked that one-frame presses and releases result in a hit correctly. Otherwise you could get intermittent misses on last objects.

Reply:
Seems ok (the objects in the last stack all have the same StartTime)

![动画](https://user-images.githubusercontent.com/50285552/126433242-2edb0dab-293b-460f-8ada-ef0ec22eef0b.gif)